beacon_node/beacon_chain/src/validator_custody.rs

```diff
@@ -0,0 +1,433 @@
+            sender,
+        }
+    }
+
+    pub fn subscribe(&self) -> Receiver<CustodyContextMessage> {
+        self.sender.subscribe()
+    }
+
+    /// Register a new validator index and updates the list of validators if required.
+    ///
+    /// Also modifies the internal structures if the validator custody has changed to
+    /// update the `custody_column_count`.
+    pub fn register_validator<E: EthSpec>(
+        &self,
+        validators_and_balance: Vec<(usize, u64)>,
+        slot: Slot,
+        spec: &ChainSpec,
+    ) {
+        // Complete all registrations
+        let mut registrations = self.validator_registrations.write();
+        for (validator_index, effective_balance) in validators_and_balance {
+            registrations.register_validator::<E>(validator_index, effective_balance, slot, spec);
+        }
+
+        // Completed registrations, now check if the validator custody requirement has changed
+        let Some(new_validator_custody) = registrations.latest_validator_custody_requirement()
+        else {
+            return;
+        };
+
+        let current_cgc = self.custody_group_count(spec);
+        let validator_custody_count_at_head =
+            self.validator_custody_count_at_head.load(Ordering::Relaxed);
+
+        if new_validator_custody != validator_custody_count_at_head {
+            tracing::debug!(
+                old_count = validator_custody_count_at_head,
+                new_count = new_validator_custody,
+                "Validator count at head updated"
+            );
+            self.validator_custody_count_at_head
+                .store(new_validator_custody, Ordering::Relaxed);
+
+            let updated_cgc = self.custody_group_count(spec);
+            // Send the message to network only if there are more columns subnets to subscribe to
+            if updated_cgc > current_cgc {
+                tracing::debug!(
+                    old_cgc = current_cgc,
+                    updated_cgc,
+                    "Custody group count updated"
+                );
+                if let Err(e) = self
+                    .sender
+                    .send(CustodyContextMessage::HeadCustodyCountChanged {
+                        new_custody_count: updated_cgc,
+                    })
+                {
+                    tracing::error!(error=?e, "Failed to send custody context message");
+                }
+            }
+        }
+    }
+
+    /// The custody count that we advertise to our peers in our metadata and
+    /// enr values.
+    pub fn advertised_custody_group_count(&self, spec: &ChainSpec) -> u64 {
+        if self.persisted_is_supernode {
+            return spec.number_of_custody_groups;
+        }
+        let advertised_validator_custody = self
+            .advertised_validator_custody_count
+            .load(Ordering::Relaxed);
+
+        // If there are no validators, return the minimum custody_requirement
+        if advertised_validator_custody > 0 {
```

Comment:
Could this be simplified to
`max(advertised_validator_custody, spec.custody_requirement)`